### PR TITLE
Basic mssql 2012 support

### PR DESCRIPTION
**Pull request by rPraml**
Hello Rob,

it would be great, if you can review the changes for sqlserver 2012.
BTW: there is a divergence between Platform and DatabasePlatform.
Especially in AnnotationBase you are trying to convert: `this.platform = Platform.valueOf(databasePlatform.getName().toUpperCase());`
So, why not using the Platform property directly?

cheers
Roland

### CODE REVIEW

File: src/main/java/com/avaje/ebean/config/dbplatform/MsSqlServer2005Platform.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
   public MsSqlServer2005Platform() {
     super();
-    this.name = "mssqlserver2005";
+    this.name = "mssqlserver";
```
Comment (rPraml) [inline src/main/java/com/avaje/ebean/config/dbplatform/MsSqlServer2005Platform.java]: ups. should be named "sqlserver"

File: pom.xml [hunk 2]
```diff
@@ -24,6 +24,34 @@
...
+				<groupId>com.microsoft.sqlserver</groupId>
+				<artifactId>sqljdbc42</artifactId>
+				<version>4.2</version>
+				<scope>test</scope>
+			</dependency>
+  		</dependencies>
+  		<repositories>
+			<repository>
+				<id>foconis</id>
+				<url>https://repo.foconis.de/maven-release/</url>
```
Comment (rPraml) [inline pom.xml]: I need the sqljdbc42 driver from microsoft which is not on mvncentral, (ok, there is one http://mvnrepository.com/artifact/org.formcept - but no official one)
so I put it to our local repository (not reachable from the internet)

File: src/main/java/com/avaje/ebean/config/dbplatform/BasicMsSqlLimiter.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+public class BasicMsSqlLimiter implements BasicSqlLimiter {
+
+  @Override
+  public String limit(String dbSql, int firstRow, int maxRows) {
+
+    StringBuilder sb = new StringBuilder(50 + dbSql.length());
+    sb.append(dbSql);
+    if (!dbSql.toLowerCase().contains("order by")) {
+      sb.append(" ORDER BY 1 ");
```
Comment (rPraml) [inline src/main/java/com/avaje/ebean/config/dbplatform/BasicMsSqlLimiter.java]: SQL-Server always needs an order by clause to support "pagination"

**Comment by 0xbaadf00d** [comment]
Is these modifications for MS-SQL 2012 only or for all MS-SQL server version >= 2012 ?

File: src/main/java/com/avaje/ebeaninternal/server/persist/dml/InsertMeta.java [hunk 4]
```diff
@@ -18,6 +19,9 @@
  */
 public final class InsertMeta {
 
+  private enum IdGeneration {
+    BY_SERVER, BY_CLIENT, BY_CLIENT_WITH_IDENTIY_INSERT;
```
Comment (rPraml) [inline src/main/java/com/avaje/ebeaninternal/server/persist/dml/InsertMeta.java]: This is the most complex part, basically there are two ways to generate an id
1. by server - when the id property is null (e.g. auto-increment/sequence)
2. by client - when the id property is set (e.g. uuid)

Unfortunately, sqlserver fails if the client tries to set/enforce an id on an auto-increment column. If you want to do this, you must prepend a "SET IDENTITY_INSERT tablename ON" before your insert statement.

So what I do here:
1. Determine if the ID of the bean is generated by the server
2. if yes - Determine, if the platform needs identity_insert
3. if yes - surround the SQL insert with "set identity_insert on/off"

**Comment by rPraml** [comment]
@0xbaadf00d it should work for versions >= 2012.  
To be clear: I don't know if it will work for 2012, I run the tests against "Microsoft SQL Server 2014 - 12.0.2269.0"
I named the platform "MsSqlServer2012" because with 2012 there came the "OFFSET/FETCH" feature to support "limit by" queries. (Before 2012 you have to write rather complex statements to get the enties "limit by 10,20" )

**Comment by rbygrave** [comment]
*> by server - when the id property is null (e.g. auto-increment/sequence)*

SQL Server does not support getting the Id value back when using JDBC batch
with auto-increment (Identity).  I suspect this is not going to change
anytime soon.

With that in mind I believe the best approach (well, the one I'd recommend)
is to use sequences and have that set via the client (via Ebean). If the
sequences are created with a decent step (say increment by 50) then this
can be very efficient with the only downside being the possibility of gaps
in the dev/testing case (where the EbeanServer is started/stopped a lot and
it grabs the next sequence and moves the value 50+ leaving gaps - it
doesn't really matter of course but this upsets some people).

This approach then matches the Oracle approach.

The only issue with adopting this approach is a decent number of existing
Ebean/SQLServer users will be using identity.


Cheers, Rob.

File: src/main/java/com/avaje/ebeaninternal/server/persist/dml/InsertMeta.java (hunk 4 above)
Comment (rbygrave) [inline src/main/java/com/avaje/ebeaninternal/server/persist/dml/InsertMeta.java]: I'm unconvinced by this part/change.  We should just adopt sequences (with a good increment by) and get full JDBC batch support.  If SqlServer ever supports Identity with getGeneratedKeys using JDBC batch then we revisit (but it hasn't happened and at this point (some years later) I suspect it won't ... so instead adopt sequences and operate like Oracle for JDBC batching inserts).

**Pull request closed without merging**